### PR TITLE
Establish codec registry and AVC registration.

### LINKED ISSUE #126: Spec better codec strings for wav ulaw / alaw
wav files are described with mime types like:
'audio/wav; codecs=1' 

I believe codecs=1 signals 'Microsoft PCM Format', as described in this RFC: https://www.ietf.org/rfc/rfc2361.txt

Other notable digits include 6:ulaw and 7:alaw. Chrome recognizes 1 via canPlayType() and kin. Firefox recognizes more. Haven't checked the other UAs. 

Such codecs strings aren't  very readable, particularly if used in WebCodecs without the accompanying 'audio/wav' prefix. Strawman  solution: revive the idea to have a registry of codecs strings. This will serve to document the nice familiar strings for other codecs and give us an option to set a new precedent for the wav strings (e.g. 'pcm', 'ulaw', 'alaw').

### LINKED ISSUE #143: Make registry for codec strings and decoder config.description (extradata)
Centralized place to see supported codec strings is obviously helpful. It also allows us to fix #126. Furthermore, we should describe what a valid description (extradata) would be for each codec (e.g. AVC -> avcc info) as well as whether the extradata is strictly required (e.g. AVC w/out extradata -> expect annex-b format).

### PR DESCRIPTION
**Pull request by chcunningham**
This moves codec-specific defintions / descriptions into a registry.
I've also added some text to the main document to clarify that support
for any given codec is not required.

Fixes #143. Fixes #126.

***
<a href="https://pr-preview.s3.amazonaws.com/WICG/web-codecs/pull/149.html" title="Last updated on Mar 17, 2021, 11:39 PM UTC (49baae7)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/web-codecs/149/57e2560...49baae7.html" title="Last updated on Mar 17, 2021, 11:39 PM UTC (49baae7)">Diff</a>

### CODE REVIEW

**Comment by chcunningham** [comment]
@tidoust FYI - this is what I'd like to land before the FPWD transition.

**Comment by chcunningham** [comment]
@aboba - looks like the pr-preview link at the bottom of the description only supports a preview of the main spec ([known issue](https://github.com/tobie/pr-preview/issues/18)). Here are some links to the generated HTML to assist in review. 

https://storage.googleapis.com/videostack-external/webcodecs_pr_149/index.html
https://storage.googleapis.com/videostack-external/webcodecs_pr_149/codec_registry.html
https://storage.googleapis.com/videostack-external/webcodecs_pr_149/avc_codec_registration.html

For now I've just done the AVC registration. Other registrations will be follow up PRs (not urgent, not blocking anything).

**Comment by chcunningham** [comment]
@padenot FYI

File: avc_codec_registration.src.html
Comment (chcunningham) [inline avc_codec_registration.src.html:98]: @tabatkins, the partial above is in a comment because I couldn't figure out how make bikeshed aware that this extends a definition from another file in this repository (index.src.html). My understanding from reading the documentation is that all idl definitions are automatically exported such that this should just work?

File: codec_registry.src.html
Comment (tidoust) [inline codec_registry.src.html:47]: The other document has "WEB-CODECS". In general, it is good practice to use the same name as the spec's shortname, and good practice to use the repo name as spec shortname, so "web-codecs". If you would like to change all names to "webcodecs", now would be a good time as we transition the repo and spec to the W3C org.

File: codec_registry.src.html
Comment (tidoust) [inline codec_registry.src.html:37]: Same comment on the `for` as in the other document.

**Review by tidoust — COMMENTED** [review]
See some comments inline.

I understand that the current final goal (pending evolutions of the W3C process to account for registries) is to publish the registry and the AVC registration as non-normative Working Group Notes. Is that correct?

**Review by chcunningham — COMMENTED** [review]
> I understand that the current final goal (pending evolutions of the W3C process to account for registries) is to publish the registry and the AVC registration as non-normative Working Group Notes. Is that correct?

Yes, that is correct. And we will seek to maintain these as non-normative as the W3C process evolves.

File: codec_registry.src.html
Comment (tidoust) [reply codec_registry.src.html]: I was more thinking of string match regardless of case rather than exact match. I do not think that there is a written recommendation about the caps style for references (for instance, I don't see any in the [Manual of Style](https://w3c.github.io/manual-of-style/#References)). A majority of specs use upper case for biblio refs. I personally prefer that too, but up to you!

To be clear, there are no written guidelines that I'm aware of to recommend that biblio refs should use the shortname of the spec, which should also match the repo name. Given that naming is notoriously hard, it just seems a good idea to be consistent.

Noted for the change of name. Let's drop the `-` during repo transition!

**Comment by tidoust** [comment]
> > I understand that the current final goal (pending evolutions of the W3C process to account for registries) is to publish the registry and the AVC registration as non-normative Working Group Notes. Is that correct?
> 
> Yes, that is correct. And we will seek to maintain the non-normative status of the registry and registrations as the W3C process evolves.

From a process perspective, this means that the specs need to be published separately. That is, we're going to publish three First Public Working Drafts:
1. The WebCodecs spec in itself, intended to become a Recommendation, under the shortname `webcodecs`
2. The registry spec, intended to become a WG Note, under the shortname `webcodecs-codec-registry`
3. The AVC (H.264) WebCodecs registration, intended to become a WG Note, under the shortname `webcodecs-avc-codec-registration`.

Practically speaking, tools such as PR preview and spec-prod are more or less designed for repos that contain only one spec. Splitting the specs to dedicated repositories as done for MSE (see [list of repos](https://github.com/w3c/media-source)) would simplify integration with these tools... That said, it also makes editing slightly more cumbersome, and it's certainly possible to workaround tools constraints.

Would you be open to creating dedicated repos for the registry and AVC WebCodecs registration specs? If so, two choices:
1. I can create the registry repos under w3c right away, initialize them with the appropriate content taken from this PR, and corresponding files can then be dropped from this PR before it gets merged. Good thing about this is that it means we'll keep this repo clean.
2. You'd prefer that PR to be merged before the actual repo transition. In that case, we'll just separate documents afterwards, but note that would mean we'll need to preserve the `codec_registry.html` and `avc_codec_registration.html` files in this repo, replacing them with empty HTML files with a meta redirect to their new locations (as also done in MSE).

I would do 1. and I'm happy to prepare that.

**Comment by aboba** [comment]
Question: What is the approval process for registry entries?  The document mentions posting an Issue to the github. Is the expectation that the specification editors will do the review/approval on their own? Or that they will recommend a disposition to the WG who will then make the decision?  Also, what happens if for some reason the Github repo is no longer active?  IANA registries sometimes allow a successor mailing list or WG to be designated.

**Comment by chcunningham** [comment]
> Question: What is the approval process for registry entries? The document mentions posting an Issue to the github. Is the expectation that the specification editors will do the review/approval on their own? Or that they will recommend a disposition to the WG who will then make the decision? Also, what happens if for some reason the Github repo is no longer active? IANA registries sometimes allow a successor mailing list or WG to be designated.

@tidoust @cwilso - pls weigh in on ^

**Comment by tidoust** [comment]
> Question: What is the approval process for registry entries? The document mentions posting an Issue to the github. Is the expectation that the specification editors will do the review/approval on their own? Or that they will recommend a disposition to the WG who will then make the decision? Also, what happens if for some reason the Github repo is no longer active? IANA registries sometimes allow a successor mailing list or WG to be designated.

First, we don't yet have a clear and crisp process in place for managing registries at W3C. That is the purpose of [ongoing discussions within the Process CG](https://github.com/w3c/w3process/issues?q=is%3Aissue+is%3Aopen+registries). See also the registries section of the presentation on [Process 2021 during last TPAC](https://www.w3.org/2020/10/TPAC/talk/Process2021#talk). Hopefully, a future revision of the W3C Process will create a clear mechanism for registries at W3C.

The direction that this PR is taking is that the registry will be published as a WG Note. The Media WG would thus be in charge of updating it (and not only the editors). For maintenance purpose, W3C tends to keep WGs around forever. If we ever close the group, the W3C team would be in fine responsible for the maintenance of the registry (granted, in the W3C process, this is more aimed at [maintenance of Recommendations](https://www.w3.org/2020/Process-20200915/#rec-amended) than at the maintenance of Notes...).

Another group may be created to take over the maintenance of the registry if needed. One perfect illustration is the Media WG, which took over maintenance of MSE/EME and of their companion format registries from the HTML Media Extensions WG.

I do not know whether these possibilities need to be specified in the registry spec itself. The text in the requirements section of the WebCodecs registry seems to be based on the one of the [MSE Byte Stream Format Registry](https://www.w3.org/TR/mse-byte-stream-format-registry/)).

W3C guarantees that archives of mailing-lists will be available forever. If you believe that this guarantee is crucial, it may be preferable to request that an email be sent to some W3C archived public mailing-list (as done for MSE/EME registries) rather than to request that an issue gets posted to a GitHub repository. That said, my understanding here is that the contents of the issue (codec string, common name, link to a public spec) will be integrated into the registry directly, so is there really a need for a strong guarantee with regards to availability of the issue over time? More precisely, is there a need for a stronger guarantee than the one that currently exists for all group discussions that take place in GitHub issue trackers?

File: codec_registry.src.html
Comment (chcunningham) [reply codec_registry.src.html]: Thanks. Updated all to use UPPER case.

Comment (chcunningham) [reply codec_registry.src.html]: Thanks. Updated all to use UPPER case.

**Comment by chcunningham** [comment]
> From a process perspective, this means that the specs need to be published separately. That is, we're going to publish three First Public Working Drafts:
> 
> The WebCodecs spec in itself, intended to become a Recommendation, under the shortname webcodecs
> The registry spec, intended to become a WG Note, under the shortname webcodecs-codec-registry
> The AVC (H.264) WebCodecs registration, intended to become a WG Note, under the shortname webcodecs-avc-codec-registration.

LGTM

> Practically speaking, tools such as PR preview and spec-prod are more or less designed for repos that contain only one spec.
> ...
> Would you be open to creating dedicated repos for the registry and AVC WebCodecs registration specs?

I don't love it. WebCodecs is going to have a large number of "registrations" (see TODO's in the registry table), and having a git repo for each one sounds cumbersome. 

It looks like spec-prod [intends to remedy this soon](https://github.com/w3c/spec-prod/issues/31). Reading the replies there, I conclude that this is widespread enough that the tooling should be updated. Meanwhile, am I correct that the current travis config happily supports multiple files (I think it publishes the entire out/ folder)?

Multi-spec PR preview is more of a nice-to-have since the workaround is trivial. The analogous [feature request](https://github.com/tobie/pr-preview/issues/18) has less momentum. [The maintainer's comments](https://github.com/tobie/pr-preview/issues/19#issuecomment-414052151) suggest its just a matter of staffing it. 

> The text in the requirements section of the WebCodecs registry seems to be based on the one of the MSE Byte Stream Format Registry.

That's correct. LMK if you find this in any way lacking. 

> More precisely, is there a need for a stronger guarantee than the one that currently exists for all group discussions that take place in GitHub issue trackers?

I think GitHub should be fine.

**Comment by chcunningham** [comment]
@aboba sounds like the text around process for registry updates is reasonable. Any other concerns that block approval?

**Review by aboba — APPROVED** [review]

**Comment by chcunningham** [comment]
Thanks everyone!

**Pull request merged**
